### ISSUE TITLE
Conan 2 Docs: Tutorial

### ISSUE THREAD

**Issue by jason-m-reich** [conan-io/docs#3202]
Hi,

Short version:
How to find Conan 2 docs for creating a cache for the dependencies of a project that is just like the project from [the first part of the Conan 2 tutorial](https://docs.conan.io/2/tutorial/consuming_packages.html) (assume there are many third party dependencies associated with the project rather than just zlib.

Long version:
I'm fairly new to using package managers with C++ and completely new to Conan. Naively, I just assume that it's natural to use Conan for these two goals (which I believe are named consuming packages and creating packages, respectively):

1. Making it easy to get all of my project's dependencies downloaded, compiled, and wired up to my project (I'm using CMake w/ presets and [the first part of the Conan 2 tutorial](https://docs.conan.io/2/tutorial/consuming_packages.html) helped me set this up already).
2. Using caching to make the dependencies from #1 available in binary form for anyone to use for quickly building my project on their local machine to help with developing my project.

The reason I'm posting this query is because [the first part of the Conan 2 tutorial](https://docs.conan.io/2/tutorial/consuming_packages.html) is completely analogous to my project (i.e. I have homemade .cpp and .h files making up my project, and dependencies just like zlib that the project depends on). But when I went to [the tutorial on creating packages](https://docs.conan.io/2/tutorial/creating_packages.html) I was thinking one or more third party dependencies would be added in addition to zlib and then instructions would be given on how to cache all of those third party dependencies: Is there a place I can go to learn how to do just that? Circling back to my expectations being naïve, am I presenting a normative use case for Conan?

**Comment by memsharded** [comment]
Hi @jason-m-reich 

Thanks for your question.

The Conan packages are always installed in the cache. It is expected, for users wanting to use Conan to install the packages, that the developers will do ``conan install`` to populate the cache with the dependencies, so it is not necessary to do an extra copy, or to commit the binaries to a different place, duplicating storage, and using a parallel way that could lead to outdated binaries too.

Said that, Conan 2.0 is providing a new way for users that want to create Conan-independent developer deploys, that might be partially relocatable. We have pending the documentation and a blog post, most likely for next week. In the meantime you want to have a look to the ``--deployer=full_deploy`` docs, because this is doing what you want: https://docs.conan.io/2/reference/extensions/deployers.html

There is this test that you can also check, validating the approach: https://github.com/conan-io/conan/blob/113cf5a2875ff3f9e77981aa178772090f4680f4/conans/test/functional/command/test_install_deploy.py#L133

Basically with ``--deployer=full_deploy`` a copy of the binaries is done to your local folder. This can be done for Release and Debug configuration too. The generated CMake files, from ``CMakeDeps`` and ``CMakeToolchain`` will be pointing to the local copy, not to the Conan cache.

Just a couple of warnings:
- The relocation of environment files only work fine in Windows. For Linux a ``sed`` is necessary to redefine some paths, but that is doable
- Only implemented and tested for ``CMake`` integration, other build systems will be done in the future if there is enough user demand.

**Comment by memsharded** [comment]
>  I simply use mv to put the .conan2, build, and venv folders into the "proprietary" directory which has our source code (I also mv the self-named "ConanPresets.json" file conan install creates for us, which contains references to the Conan presets files buried within the build folder).

I think this shouldn't be done, as it might be problematic. Some packages and some stuff in the cache might break because of moving the cache.

> could sed be used recursively from the root of the project to change the original path from the base image to the relevant path that is actually being used?

There are situations where file paths might be embedded in binaries. For example, for debugging it is quite important that the packages are not moved, and doing ``sed`` on binaries it most likely quite unsafe and problematic. The ``sed`` I was talking about is done only in the ``generators`` folder, to be able to retarget the location of some ``.sh`` Conan generated files (there is no way in pure shell to have scripts know their current script folder)

> setting up a Conan remote (like Artifactory CE) for storing our dependencies cache be a normal pattern and a natural alternative to storing our dependencies cache in our base Docker image?

That sounds definitely a better practice. It allows in general to be faster avoiding tons of re-building from source, to track things better, and to evolve more easily, specially when different configurations start to be relevant.

Not sure if this helps, please let me know, I might be missing a bit of the whole picture, like the use case of that docker setup (more dev-oriented, or runtime oriented? etc)

**Comment by jason-m-reich** [comment]
> That sounds definitely a better practice. It allows in general to be faster avoiding tons of re-building from source, to track things better, and to evolve more easily, specially when different configurations start to be relevant.

Okay terrific. With acknowledgements that what I'm about to say is a newbie discovery: This ☝️ seems to solve the problem "I just did the Conan installs so I have the project dependency binaries cached, but now I want the team and/or CI to have the same [without checking the binaries into git and without doing the misguided Docker hack I mentioned above]".

If what I just wrote is sane, then I'm good to go. This has helped a ton! (And yes, you definitely weren't missing anything... we are just doing dev-oriented work that includes testing builds with CI). Thanks for taking the time to point out why the Docker hack was not a good idea...

**Comment by memsharded** [comment]
Perfect, thanks for the feedback!
I'll ping in this issue when the new docs for this are there.
